logstash_async/transport.py

```diff
@@ -168,7 +168,10 @@ def close(self):
 
     # ----------------------------------------------------------------------
     def send(self, events, use_logging=False):
-        client = pylogbeat.PyLogBeatClient(use_logging=use_logging, **self._client_arguments)
-        with client:
-            for events_subset in ichunked(events, self._batch_size):
-                client.send(events_subset)
+        try:
+            client = pylogbeat.PyLogBeatClient(use_logging=use_logging, **self._client_arguments)
+            with client:
+                for events_subset in ichunked(events, self._batch_size):
+                    client.send(events_subset)
+        except (socket.gaierror, ConnectionError) as send_exception:
```

Comment:
Cool. Only a few remarks:
- catching `socket.gaierror` here is a bit out of scope as these are usually name resolution errors; `ConnectionError` is fine as it covers most usual errors like 'connection refused'  and 'broken pipe'
- additionally we should catch `TimeoutError` as this is the other main cause for transmission problems (`TimeoutError` is not a sub class of `ConnectionError`)

- here is the wrong place to handle this error, especially when only printing the exception instead of logging it.
A more appropriate place is https://github.com/eht16/python-logstash-async/blob/master/logstash_async/worker.py#L214: add a new except block for `TimeoutError` and `ConnectionError` and then log the exception without stacktrace

Reply:
I don't see where else it makes sense to catch a socket error; the socket call is being made in transport.py.  I notify the user that it is an "irrecoverable error", as there isn't anything that will be done in python-logstash-async that's going to fix a socket.gaierror (i.e. someone needs to fix something outside the scope of this library).

I moved the ConnectionError (and added TimeoutError) to where you suggested in worker.py and tried to make it fit the look and feel of your code.

Reply:
> I don't see where else it makes sense to catch a socket error; the socket call is being made in transport.py. I notify the user that it is an "irrecoverable error", as there isn't anything that will be done in python-logstash-async that's going to fix a socket.gaierror (i.e. someone needs to fix something outside the scope of this library).

I don't mind so much to keep `socket-gaierror`. It's just different kind of causes than `ConnectionError` and `Timeout` which are quite common. `socket.gaierror` are mainly name resolution errors (i.e. DNS problems). But still I'm OK with catching it.

But my main point here was that catching it here will only affect the `BeatsTransport` but the other transports could raise the same errors as well. And this is why I suggested to move *all* the error handling of such errors to `worker.py` as this is the place where all transports are used.

So, if you could add the `socket.gaierror` to the exception list in `worker.py` and get rid of that `print()` statement, it would do what you want for `BeatsTransport` and what I want (for all other transports as well).

> I moved the ConnectionError (and added TimeoutError) to where you suggested in worker.py and tried to make it fit the look and feel of your code.

Great, thanks.

Reply:
The socket.gaierror exception has been moved to worker.py and made similar to other exceptions in the list.

Reply:
Cool.
I don't understand yet why we handle `gaierror` differently from `ConnectionError` and `Timeout`, I would consider them to have the same log level (ERROR) and so be handled in the same `except` block.
IMO a `gairerror` is in the same way out of scope of the library as `ConnectionError` and `Timeout` are.
What do you think?

Reply:
It was mostly because you didn't want those exceptions grouped in the original place I put them.

This isn't my project, and I need to do these things to *your* standards to get them included upstream so I don't have to make and maintain a custom build of your library.  If you want me to group the exceptions all as one, then that's what I'll do.

Reply:
My initial point about `gaierror` was that I don't consider it as so common to not have a stacktrace for them. But it's probably also fine to do so as you did.
Sorry for not being clear enough in the first place.

Thanks for the changes!